### PR TITLE
better features overview

### LINKED ISSUE #17733: Improve feature (detection) code in the Admin Console
There is already [some code](https://github.com/keycloak/keycloak/blob/aa77f713c48fb2f0678ffd4cc2318c1c4c400b79/js/apps/admin-ui/src/utils/useIsFeatureEnabled.ts) present to detect which features are enabled in the Admin Console. However, due to the way the REST API is designed it cannot properly detect all of the enabled features. Therefore this implementation must be changed.

The proposal is as follows, instead of the [`ProfileInfoRepresentation`](https://www.keycloak.org/docs-api/21.0.1/rest-api/index.html#_profileinforepresentation) containing a bunch of separate fields (`disabledFeatures`, `experimentalFeatures` and `previewFeatures`) it should contain a single `features` field with a more detailed data structure that matches what is known in the model on the Java side.

An example of such a data structure would be:
```js
{
  features: [
    {
      id: "AUTHORIZATION",
      label: "Authorization Service",
      type: "DEFAULT",
      isEnabled: true,
      dependencies: []
    },
    {
      id: "ACCOUNT_API",
      label: "Account Management REST API",
      type: "DEFAULT",
      isEnabled: true,
      dependencies: []
    },
    {
      id: "ACCOUNT3",
      label: "Account Management Console version 3",
      type: "EXPERIMENTAL",
      isEnabled: false,
      dependencies: ["ACCOUNT_API"]
    }
  ]
}
```

Then, once the REST API has been adjusted the data can be retrieved and mapped into something like the following TypeScript:

```ts
enum FeatureType {
  /** This feature is enabled by default. */
  Default = "DEFAULT",
  /** This feature is disabled by default. */
  DisabledByDefault = "DISABLED_BY_DEFAULT",
  /** This feature is in preview, it will be stabilized soon and is enabled by default. */
  Preview = "PREVIEW",
  /** This feature is in preview, it will be stabilized soon but is disabled by default. */
  PreviewDisabledByDefault = "PREVIEW_DISABLED_BY_DEFAULT",
  /** This feature is experimental and disabled by default, it might be promoted to preview in the future. */
  Experimental = "EXPERIMENTAL",
  /** This feature is deprecated and disabled by default, it should not be used and will be removed in the future. */
  Deprecated = "DEPRECATED"
}

type Feature = {
  /** The unique identifier for the feature. */
  id: string;
  /** The type of feature. */
  type: FeatureType;
  /** A human readable label that describes the feature. */
  label: string;
  /** A boolean indicating whether the feature is currently enabled. */
  isEnabled: boolean;
  /** Dependencies of the feature, which need to be enabled in order for it to work. */
  dependencies: Feature[];
}
```

Subsequently we can then refactor the code that previously used the now deprecated `disabledFeatures`, `experimentalFeatures` and `previewFeatures` fields to the new `features` field, which should contain accurate information about whether which features are enabled.

### PR DESCRIPTION
**Pull request by edewit**
fixes: #17733

### CODE REVIEW

File: js/apps/admin-ui/src/dashboard/Dashboard.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/dashboard/Dashboard.tsx:76]: I assume we want things to on the same line, so we could also use a non breaking space here.
```suggestion
      {feature.name}&nbsp;
```

File: js/apps/admin-ui/src/dashboard/Dashboard.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/dashboard/Dashboard.tsx:72]: Can we move this into a seperate type?
```suggestion
type FeatureItemProps = {
  feature: FeatureRepresentation
};

const FeatureItem = ({ feature }: FeatureItemProps) => {
```

File: js/apps/admin-ui/src/dashboard/Dashboard.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/dashboard/Dashboard.tsx:75]: There should be no need to specify the `key` here, since that is passed when this component is used.
```suggestion
    <ListItem className="pf-u-mb-sm">
```

File: js/apps/admin-ui/src/dashboard/Dashboard.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/dashboard/Dashboard.tsx:85]: There should be no need to use a ternary expression here:
```suggestion
      {feature.type === FeatureType.EXPERIMENTAL && (
        <Label color="orange">{t("experimental")}</Label>
      )}
      {feature.type === FeatureType.PREVIEW && (
        <Label color="blue">{t("preview")}</Label>
      )}
      {feature.type === FeatureType.DEFAULT && (
        <Label color="green">{t("supported")}</Label>
      )}
```

File: js/apps/admin-ui/src/dashboard/Dashboard.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/dashboard/Dashboard.tsx:112]: I think we can reduce the amount of repetition here by first sorting the features based on the locale, and then using that to split out the enabled and disabled features:

```ts
const sortedFeatures = useMemo(
  () => localeSort(serverInfo.features ?? [], mapByKey("name")),
  [serverInfo.features],
);

const disabledFeatures = useMemo(
  () => sortedFeatures.filter(({ enabled }) => !enabled),
  [sortedFeatures],
);

const enabledFeatures = useMemo(
  () => sortedFeatures.filter(({ enabled }) => enabled),
  [sortedFeatures],
);
```

Also note, that even though the linter complains, we should not add `localeSort` to the dependency array. As this would cause the memo to-retrigger every render.

File: js/apps/admin-ui/src/utils/useIsFeatureEnabled.ts
Comment (jonkoops) [inline js/apps/admin-ui/src/utils/useIsFeatureEnabled.ts:19]: This code doesn't always return a `boolean`, if `features` is not defined it will return `undefined`. We can add an early return to fix this:
```suggestion
    if (!features) {
      return false;
    }

    return features
      .filter((f) => f.enabled)
      .map((f) => f.name)
      .includes(feature);
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: js/libs/keycloak-admin-client/src/defs/featureRepresentation.ts
Comment (edewit) [reply js/libs/keycloak-admin-client/src/defs/featureRepresentation.ts:15]: actually this is the convention

Comment (jonkoops) [reply js/libs/keycloak-admin-client/src/defs/featureRepresentation.ts:15]: This is what we're doing in some places yet, but as per the [TypeScript Handbook](https://www.typescriptlang.org/docs/handbook/enums.html) the convention is to use pascal case.

File: js/apps/admin-ui/src/dashboard/Dashboard.tsx
Comment (jonkoops) [inline js/apps/admin-ui/src/dashboard/Dashboard.tsx:77]: I think we can omit the namespace here now right?
```suggestion
  const { t } = useTranslation();
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Review by jonkoops — APPROVED** [review]

**Comment by jonkoops** [comment]
@ssilvert can you review this one as well? It needs maintainer approval.

**Review by ssilvert — APPROVED** [review]
LGTM

**Pull request merged**
